src/v/serde/parquet/shredder.cc

```diff
@@ -0,0 +1,279 @@
+          // If the value is `null`, we use the parent definition_level so that
+          // assembly can determine where the `null` started.
+          for (const auto& child : element.children) {
+              cb(child, levels, value(v));
+          }
+      },
+      [&element](list_value&) {
+          throw std::runtime_error(fmt::format(
+            "unexpected list value for non-repeated schema element {}",
+            element.name));
+      },
+      [&element](map_value&) {
+          throw std::runtime_error(fmt::format(
+            "unexpected map value for non-repeated schema element {}",
+            element.name));
+      },
+      [&element](auto& v) {
+          throw std::runtime_error(fmt::format(
+            "unexpected leaf value for optional schema element {}: {}",
+            element.name,
+            value(std::move(v))));
+      });
+}
+
+void for_each_repeated_child(
+  value val,
+  const indexed_schema_element& element,
+  traversal_levels levels,
+  absl::FunctionRef<
+    void(const indexed_schema_element&, traversal_levels, value)> cb) {
+    ss::visit(
+      std::move(val),
+      [&element, levels, cb](null_value& v) {
+          for_each_optional_child(v, element, levels, cb);
+      },
+      [&element, levels, cb](list_value& list) {
+          // Empty lists are equivalent to a `null` value.
+          if (list.empty()) {
+              for_each_optional_child(null_value(), element, levels, cb);
+              return;
+          }
+          traversal_levels child_levels = levels;
+          // We are marking that there is a higher depth here we
+          // need to record, but the repetition_level is only set
+          // *when* we are repeating (so not for the first element).
+          ++child_levels.repetition_depth;
+          for (auto& member : list) {
+              for_each_optional_child(
+                std::move(member.element), element, child_levels, cb);
+              child_levels.repetition_level = child_levels.repetition_depth;
+          }
+      },
+      [&element](map_value&) {
+          throw std::runtime_error(
+            fmt::format("TODO(rockwood) handle maps: {}", element.name));
+      },
+      [&element](struct_value&) {
+          throw std::runtime_error(fmt::format(
+            "unexpected struct value for repeated schema element {}",
+            element.name));
+      },
+      [&element](auto& v) {
+          throw std::runtime_error(fmt::format(
+            "unexpected leaf value for repeated schema element {}: {}",
+            element.name,
+            value(std::move(v))));
+      });
+}
+
+void for_each_child(
+  value val,
+  const indexed_schema_element& element,
+  traversal_levels levels,
+  absl::FunctionRef<
+    void(const indexed_schema_element&, traversal_levels, value)> cb) {
+    switch (element.repetition_type) {
+    case field_repetition_type::required:
+        for_each_required_child(std::move(val), element, levels, cb);
```

Comment:
I took a stab at renaming all these functions - TBH I don't really like the `for_each_child` naming anyways. Really we're processing an inner node in the schema tree. I changed the naming to be split to when we process a schema node vs a value node since we're traversing two trees in tandem. I added some color to the commit message about the reasoning behind the naming.